### PR TITLE
Enable `TagHelper`s to properly flatten to enable partial parsing.

### PR DESCRIPTION
**Pull request by NTaylorMullen**
- `TagHelper`s used to not flatten correctly resulting in inconsistent start locations for `SyntaxTreeNode`s following/within `TagHelper`s.
- This change indirectly corrected bad indentation that existed in generated C#.
- Added a test to validate `TreesAreDifferent` returns expected behavior when modifying content inside of `TagHelper`.
- Added `TagHelperBlock` `Flatten` test.

#553

### CODE REVIEW

File: src/Microsoft.AspNet.Razor/Parser/TagHelpers/TagHelperBlockBuilder.cs [hunk 1]
```diff
@@ -134,6 +134,6 @@ public override void Reset()
         /// <summary>
         /// The starting <see cref="SourceLocation"/> of the tag helper.
         /// </summary>
-        public SourceLocation Start { get; private set; }
+        public SourceLocation Start { get; set; }
```
Comment (ajaybhargavb) [inline src/Microsoft.AspNet.Razor/Parser/TagHelpers/TagHelperBlockBuilder.cs]: Is this change needed only for testing? If so shouldn't this just be `internal`?

Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Razor/Parser/TagHelpers/TagHelperBlockBuilder.cs]: In this PR we need it accessible for testing yes. However, the builder was a bit odd to begin with only having a `private set`. All of our other builds have public get/set methods, thought I'd just normalize this property since I had to touch it anyways.

Comment (ajaybhargavb) [inline src/Microsoft.AspNet.Razor/Parser/TagHelpers/TagHelperBlockBuilder.cs]: Got it.

**Comment by ajaybhargavb** [comment]
:shipit:

**Comment by dougbu** [comment]
:shipit:
